test/integration/consul-container/test/README.md

```diff
@@ -0,0 +1,123 @@
+| Tags / Flags      | Default value | Description |
+| -----------         | ----------- | ----------- |
+| --latest-image      | consul    | Refers to the docker image name to be deployed before upgrade
+| --latest-version    | latest      | Refers to the version or tag of latest image to be deployed before upgrade
+| --target-image      | consul    | Refers to the image name that will be upgrade to
+| --target-version    | local     | Refers to the version of target image to upgrade to
+| -follow-log         |  false    | This optional flag emit event logs
+
+
+
+## Adding a new upgrade integration test  
+Upgrade integration tests are defined in the [test/integration/consul-container/test/upgrade](/test/integration/consul-container/test/upgrade) subdirectory and new upgrade integration tests should always be added to this location. The test framework uses
+[functional table-driven tests in Go](https://yourbasic.org/golang/table-driven-unit-test/) and using function types to modify the base value for each test case.
+
+For tests with multiple test cases, it should always start by invoking 
+```go
+	type testcase struct {
+		name string
+		create func()
+		extraAssertion func()
+	}
+```
+see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/l7_traffic_management/resolver_default_subset_test.go). For upgrade tests with a single test case, they can be written like
+```go
+	run := func(t *testing.T, oldVersion, targetVersion string) {
+        // insert test
+    }
+	t.Run(fmt.Sprintf("Upgrade from %s to %s", utils.LatestVersion, utils.TargetVersion),
+		func(t *testing.T) {
+			run(t, utils.LatestVersion, utils.TargetVersion)
+		})
+```
+see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/acl_node_test.go)
+
+### How it works
+![Upgrade Tests Workflow](util/upgrade_tests_workflow.png?raw=true)
+
+A Consul cluster is deployed, then a static-server, static-client and envoy sidecars are created in the cluster. An API request is made to the static-client to validate that it is ready. 
+
+Then we validate traffic between the static-server and static-client envoy sidecar. After validation, we take a snapshot and upgrade the Consul cluster to the `target-version` and restart the sidecars. Re-validate the client, server and sidecars to ensure the data snapshotted from the pervious version can be accessed in the latest version. 
+
+
+### Errors Test Cases
+There are some caveats for special error handling of versions prior to `1.14`.   
+Upgrade tests for features such peering, had API changes that returns an error if attempt to upgrade, and should be accounted for in upgrade tests. If running upgrade tests for any version before `1.14`, the following lines of code needs to be added to skip test or it will not pass.  
+
+```go
+		fromVersion, err := version.NewVersion(utils.LatestVersion)
+		require.NoError(t, err)
+		if fromVersion.LessThan(utils.Version_1_14) {
+			continue
```

Comment:
I like to use e,g, `t.Skipf("version %s (< 1.14) doesn't support <blah>", utils.LatestVersion)`